### ISSUE TITLE
[Phase 3] Temporal Cross-Validation

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/demand-forecasting-engine#4]
## Summary

Implement proper time series cross-validation with expanding windows to get robust performance estimates. This replaces single train/test split with multiple temporal folds for reliable model evaluation.

## System Context

```
demand-forecasting-engine/
├── src/
│   └── evaluation/
│       ├── __init__.py
│       ├── metrics.py         # EXISTS
│       ├── splits.py          # EXISTS
│       ├── cv.py              # <- CREATE
│       └── splitters.py       # <- CREATE
└── scripts/
    └── run_cv.py              # <- CREATE
```

## Dependencies

- Requires: Phase 2 (Single Model Baseline)
- Blocks: Phase 4 (Quantile Regression), Phase 5 (Ensemble)

### 1. Custom Time Series Splitter (`src/evaluation/splitters.py`)

```python
import numpy as np
import pandas as pd
from typing import Iterator, Tuple, Optional


class ExpandingWindowSplitter:
    """
    Expanding window cross-validation for time series.

    The training set grows with each fold while test set moves forward.

    Example with 5 folds:
    Fold 1: [████████] train | gap | [██] test
    Fold 2: [████████████] train | gap | [██] test
    Fold 3: [████████████████] train | gap | [██] test
    Fold 4: [████████████████████] train | gap | [██] test
    Fold 5: [████████████████████████] train | gap | [██] test

    Parameters
    ----------
    n_splits : int
        Number of CV folds
    test_size : int
        Number of samples in each test fold
    gap : int
        Number of samples to skip between train and test (prevents leakage)
    min_train_size : int, optional
        Minimum samples required in training set
    """

    def __init__(
        self,
        n_splits: int = 5,
        test_size: int = 28,
        gap: int = 7,
        min_train_size: Optional[int] = None,
    ):
        self.n_splits = n_splits
        self.test_size = test_size
        self.gap = gap
        self.min_train_size = min_train_size

    def split(
        self,
        X: pd.DataFrame,
        y: Optional[pd.Series] = None,
        groups: Optional[pd.Series] = None,
    ) -> Iterator[Tuple[np.ndarray, np.ndarray]]:
        """
        Generate train/test indices for each fold.

        Parameters
        ----------
        X : pd.DataFrame
            Feature matrix (used for length only)
        y : pd.Series, optional
            Target (not used, for sklearn compatibility)
        groups : pd.Series, optional
            Group labels (not used)

        Yields
        ------
        train_indices, test_indices : tuple of np.ndarray
        """
        n_samples = len(X)

        # Calculate fold boundaries
        total_test = self.n_splits * self.test_size
        total_gap = self.n_splits * self.gap

        if self.min_train_size:
            min_train = self.min_train_size
        else:
            min_train = (n_samples - total_test - total_gap) // 2

        # Generate folds
        for fold in range(self.n_splits):
            test_end = n_samples - (self.n_splits - fold - 1) * (self.test_size + self.gap)
            test_start = test_end - self.test_size
            train_end = test_start - self.gap

            train_indices = np.arange(0, train_end)
            test_indices = np.arange(test_start, test_end)

            if len(train_indices) >= min_train:
                yield train_indices, test_indices

    def get_n_splits(self, X=None, y=None, groups=None) -> int:
        """Return the number of splits."""
        return self.n_splits


class SlidingWindowSplitter:
    """
    Sliding window cross-validation (fixed training size).

    Training window slides forward with each fold.

    Example with 3 folds:
    Fold 1: [████████] train | gap | [██] test
    Fold 2:     [████████] train | gap | [██] test
    Fold 3:         [████████] train | gap | [██] test

    Parameters
    ----------
    n_splits : int
        Number of CV folds
    train_size : int
        Fixed size of training window
    test_size : int
        Number of samples in each test fold
    gap : int
        Gap between train and test
    step_size : int
        How much to slide the window each fold
    """

    def __init__(
        self,
        n_splits: int = 5,
        train_size: int = 365,
        test_size: int = 28,
        gap: int = 7,
        step_size: Optional[int] = None,
    ):
        self.n_splits = n_splits
        self.train_size = train_size
        self.test_size = test_size
        self.gap = gap
        self.step_size = step_size or test_size

    def split(
        self,
        X: pd.DataFrame,
        y: Optional[pd.Series] = None,
        groups: Optional[pd.Series] = None,
    ) -> Iterator[Tuple[np.ndarray, np.ndarray]]:
        """Generate train/test indices."""
        n_samples = len(X)

        for fold in range(self.n_splits):
            train_start = fold * self.step_size
            train_end = train_start + self.train_size
            test_start = train_end + self.gap
            test_end = test_start + self.test_size

            if test_end > n_samples:
                break

            train_indices = np.arange(train_start, train_end)
            test_indices = np.arange(test_start, test_end)

            yield train_indices, test_indices

    def get_n_splits(self, X=None, y=None, groups=None) -> int:
        return self.n_splits
```

### 2. Cross-Validation Runner (`src/evaluation/cv.py`)

```python
import numpy as np
import pandas as pd
from typing import Dict, List, Any, Optional
from tqdm import tqdm

from .splitters import ExpandingWindowSplitter
from .metrics import calculate_metrics
from ..models.base_model import BaseForecaster


def temporal_cross_validate(
    model: BaseForecaster,
    X: pd.DataFrame,
    y: pd.Series,
    n_splits: int = 5,
    test_size: int = 28,
    gap: int = 7,
    splitter_type: str = 'expanding',
    return_models: bool = False,
    verbose: bool = True,
) -> Dict[str, Any]:
    """
    Perform temporal cross-validation.

    Parameters
    ----------
    model : BaseForecaster
        Model to evaluate (will be cloned for each fold)
    X : pd.DataFrame
        Feature matrix
    y : pd.Series
        Target values
    n_splits : int
        Number of CV folds
    test_size : int
        Number of samples in each test fold
    gap : int
        Gap between train and test to prevent leakage
    splitter_type : str
        'expanding' or 'sliding'
    return_models : bool
        Whether to return trained models from each fold
    verbose : bool
        Whether to show progress bar

    Returns
    -------
    dict
        {
            'mae': np.ndarray of MAE per fold,
            'rmse': np.ndarray of RMSE per fold,
            'mape': np.ndarray of MAPE per fold,
            'smape': np.ndarray of SMAPE per fold,
            'fold_info': List of fold metadata,
            'models': List of models (if return_models=True),
        }
    """
    if splitter_type == 'expanding':
        splitter = ExpandingWindowSplitter(n_splits, test_size, gap)
    else:
        splitter = SlidingWindowSplitter(n_splits, test_size=test_size, gap=gap)

    results = {
        'mae': [],
        'rmse': [],
        'mape': [],
        'smape': [],
        'fold_info': [],
    }
    if return_models:
        results['models'] = []

    splits = list(splitter.split(X))
    iterator = tqdm(splits, desc="CV Folds") if verbose else splits

    for fold_idx, (train_idx, test_idx) in enumerate(iterator):
        X_train, X_test = X.iloc[train_idx], X.iloc[test_idx]
        y_train, y_test = y.iloc[train_idx], y.iloc[test_idx]

        # Clone model for this fold
        fold_model = model.__class__(**model.get_params())

        # Train
        fold_model.fit(X_train, y_train)

        # Predict
        y_pred = fold_model.predict(X_test)

        # Calculate metrics
        metrics = calculate_metrics(y_test.values, y_pred)

        for metric_name, metric_value in metrics.items():
            results[metric_name].append(metric_value)

        results['fold_info'].append({
            'fold': fold_idx,
            'train_size': len(train_idx),
            'test_size': len(test_idx),
        })

        if return_models:
            results['models'].append(fold_model)

    # Convert to numpy arrays
    for metric in ['mae', 'rmse', 'mape', 'smape']:
        results[metric] = np.array(results[metric])

    return results


def print_cv_results(results: Dict[str, Any]) -> None:
    """Pretty print cross-validation results."""
    print("\nCross-Validation Results")
    print("=" * 50)

    for metric in ['mae', 'rmse', 'mape', 'smape']:
        values = results[metric]
        print(f"{metric.upper():6s}: {values.mean():.4f} (+/- {values.std():.4f})")

    print("\nPer-Fold Details:")
    print("-" * 50)
    for i, info in enumerate(results['fold_info']):
        print(f"Fold {i+1}: train={info['train_size']:,}, test={info['test_size']:,}, "
              f"MAPE={results['mape'][i]:.2%}")
```

### 3. CV Runner Script (`scripts/run_cv.py`)

```python
#!/usr/bin/env python
"""Run temporal cross-validation."""
import argparse
import pandas as pd

from src.data.loader import load_sales_data
from src.features.pipeline import build_feature_pipeline
from src.models.lightgbm_model import LightGBMForecaster
from src.evaluation.cv import temporal_cross_validate, print_cv_results


def main():
    parser = argparse.ArgumentParser(description="Run cross-validation")
    parser.add_argument("--data", required=True, help="Path to sales data")
    parser.add_argument("--n-splits", type=int, default=5, help="Number of CV folds")
    parser.add_argument("--test-size", type=int, default=28, help="Test size per fold")
    parser.add_argument("--gap", type=int, default=7, help="Gap between train/test")
    args = parser.parse_args()

    # Load and prepare data
    print("Loading data...")
    df = load_sales_data(args.data)

    print("Building features...")
    pipeline = build_feature_pipeline()
    df_features = pipeline.fit_transform(df)

    feature_cols = [c for c in df_features.columns
                    if c not in ['date', 'sku_id', 'sales']]
    X = df_features[feature_cols].dropna()
    y = df_features.loc[X.index, 'sales']

    # Run CV
    print(f"\nRunning {args.n_splits}-fold CV...")
    model = LightGBMForecaster()
    results = temporal_cross_validate(
        model=model,
        X=X,
        y=y,
        n_splits=args.n_splits,
        test_size=args.test_size,
        gap=args.gap,
        verbose=True,
    )

    print_cv_results(results)


if __name__ == "__main__":
    main()
```

## Implementation Checklist

- [ ] Implement `ExpandingWindowSplitter` with gap support
- [ ] Implement `SlidingWindowSplitter` as alternative
- [ ] Create `temporal_cross_validate()` function
- [ ] Add per-fold metric aggregation
- [ ] Implement model cloning for independent fold training
- [ ] Add progress bar for CV execution
- [ ] Create `run_cv.py` script with CLI
- [ ] Visualize train/test windows

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/evaluation/splitters.py` | 0-120 | Create | Custom CV splitters |
| `src/evaluation/cv.py` | 0-100 | Create | CV runner function |
| `scripts/run_cv.py` | 0-50 | Create | CLI script |

## Technical Challenges

**Understanding the Gap Parameter**: The gap prevents data leakage from features using recent data.

```
Without gap (WRONG):
Train: [...day 98, day 99, day 100] | Test: [day 101, day 102...]
Problem: lag_1 feature on day 101 uses day 100 (in training)

With gap=7 (CORRECT):
Train: [...day 93, day 94] | gap [95-101] | Test: [day 102, 103...]
Now: lag_7 on day 102 uses day 95, which is excluded from train
```

## Definition of Done

- [ ] `ExpandingWindowSplitter.split()` generates correct indices
- [ ] Gap parameter correctly excludes buffer zone
- [ ] CV returns per-fold and aggregate metrics
- [ ] No data leakage between folds
- [ ] Tests pass: `pytest tests/test_cv.py -v`

## Risk Assessment

| Risk | Impact | Mitigation |
|------|--------|------------|
| Folds overlapping | :red_circle: High | Unit test index boundaries |
| Inconsistent fold sizes | :yellow_circle: Medium | Log fold info |
| Slow CV with large data | :yellow_circle: Medium | Add progress bar |

## Time Estimate

**Estimated Hours**: 8-10 hours
**Priority**: High (enables robust evaluation)
